libsrc/_DEVELOPMENT/math/float/math32/z80/f32_z80_mulu_de.asm

```diff
@@ -0,0 +1,117 @@
+;
+;  feilipu, 2019 May
+;
+;  This Source Code Form is subject to the terms of the Mozilla Public
+;  License, v. 2.0. If a copy of the MPL was not distributed with this
+;  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+;
+;-------------------------------------------------------------------------
+;
+; REPLICATION for Z80 of:
+; Z180 MLT DE and Z80-ZXN MUL DE
+;
+; unsigned multiplication of two 8-bit
+; multiplicands into a sixteen bit product
+;
+; enter : d = 8-bit multiplicand
+;         e = 8-bit multiplicand
+;
+; exit  : de = 16-bit product
+;         carry reset
+;
+; uses  : f, de
+   
+SECTION code_clib
+SECTION code_math
+
+PUBLIC m32_z80_mulu_de
+
+
+.m32_z80_mulu_de
+
+   inc d
+   dec d
+   jr Z,zero                ; multiply by 0
+
+   inc e
+   dec e
+   jr Z,zero                ; multiply by 0
+
+   push hl
+   ex de,hl
+
+   ld e,l
+   ld d,0
+
+   ; eliminate leading zero bits
+
+   sla h
+   jr C,branch_11
+
+   sla h
+   jr C,branch_12
+
+   sla h
+   jr C,branch_13
+
+   sla h
+   jr C,branch_14
+
+   sla h
+   jr C,branch_15
+
+   sla h
+   jr C,branch_16
+
+   sla h
+   jr C,branch_17
+
+   sla h
+   jr exit                ; multiply by 1
+   
+   ; multiplication tree
+
+.branch_11
+   add hl,hl
+   jr NC,branch_12
+   add hl,de
+
+.branch_12
+   add hl,hl
+   jr NC,branch_13
+   add hl,de
+
+.branch_13
+   add hl,hl
+   jr NC,branch_14
+   add hl,de
+
+.branch_14
+   add hl,hl
+   jr NC,branch_15
+   add hl,de
+
+.branch_15
+   add hl,hl
+   jr NC,branch_16
+   add hl,de
+
+.branch_16
+   add hl,hl
+   jr NC,branch_17
+   add hl,de
+
+.branch_17
+   add hl,hl
+   jr NC,exit
+   add hl,de
+
+.exit
+   ex de,hl
+   pop hl
+   or a
+   ret
+
+.zero
```

Comment:
to skim few more cycles, this can be broken into two tails, `zeroD` and `zeroE` (depending which one was zero), each of them doing either `ld e,d` or `ld d,e` + `ret`. ... now I even actually wonder how often you multiply by zero, if the initial test is worth it, as the `sla h` for leading zero bits takes roughly about 8 x (8+7) = 120T extra for zero input, but for non zero input the initial test is 30T, i.e. for 65025 out of 65536 cases there's +30T extra, remaining 511 cases save about 120T => i.e. roughly1950750T lost (non zero input) vs 61320T gained (when zero input). ... seems like I missed something obvious, and my reasoning is wrong somewhere, or you have hard-data from real life suggesting mul by zero is used a lot (but if then also mul-by-1 is a common thing, you can make that super fast by changing entry to `dec d / ret z / inc d / jr Z,zero`).

Reply:
Yes, I agree. Actually I woke up thinking the same thing this morning, about splitting the tails.
Done.

But, more generally about whether zero detection is worth it or not? I've no idea, actually.

My thinking was coloured by two things. First, when initially using the rabbit style logic, there were a lot of dead multiplies, where one byte of the 16 bit pair would be set to zero. Implementing this on the z80 (z180, z80-zxn) meant a lot of needless multiplication. But, I've resolved this now, and it won't be an issue.

Second, a lot of integer like numbers have "short" mantissas. If we call a 24 bit mantissa made up of three bytes "abc", then quite often the b and c bytes end up being zero. I think this will be a common case. Handling these with zero detection will be a big win, making many multiplies very fast (because the result is zero). Of course the other side of the argument is that integers should be handled by the integer library, and that calculating a 10 iteration polynomial estimation for `logf()` (for example) will never have zeros in it, so carrying zero detection overhead for short mantissa floats is just wasteful.

I guess the only way to find out is to benchmark this. Another box to tick.

Just to comment, I don't see any need to specially detect a 1 byte. That will come forward no more often than any other value.